internal/controllers/virtual_devices_controller.go

```diff
@@ -0,0 +1,210 @@
+}
+
+type SignVirtualDeviceMintingPayloadRequest struct {
+	VehicleNode int `json:"vehicleNode"`
+	Credentials struct {
+		AuthorizationCode string `json:"authorizationCode"`
+	} `json:"credentials"`
+	OwnerSignature string `json:"ownerSignature"`
+}
+
+func NewVirtualDeviceController(
+	settings *config.Settings, dbs func() *db.ReaderWriter, logger *zerolog.Logger, integSvc services.DeviceDefinitionIntegrationService, deviceDefSvc services.DeviceDefinitionService, usersClient pb.UserServiceClient,
+) VirtualDeviceController {
+	return VirtualDeviceController{
+		Settings:     settings,
+		DBS:          dbs,
+		log:          logger,
+		integSvc:     integSvc,
+		usersClient:  usersClient,
+		deviceDefSvc: deviceDefSvc,
+	}
+}
+
+func (vc *VirtualDeviceController) getVirtualDeviceMintPayload(integrationID int64, vehicleNode int64) *signer.TypedData {
+	return &signer.TypedData{
+		Types: signer.Types{
+			"EIP712Domain": []signer.Type{
+				{Name: "name", Type: "string"},
+				{Name: "version", Type: "string"},
+				{Name: "chainId", Type: "uint256"},
+				{Name: "verifyingContract", Type: "address"},
+			},
+			"MintDevice": {
+				{Name: "integrationNode", Type: "uint256"},
+				{Name: "vehicleNode", Type: "uint256"},
+			},
+		},
+		PrimaryType: "MintVirtualDeviceSign",
+		Domain: signer.TypedDataDomain{
+			Name:              "DIMO",
+			Version:           "1",
+			ChainId:           math.NewHexOrDecimal256(vc.Settings.DeviceMintingChainID),
+			VerifyingContract: vc.Settings.DeviceMintingVerifyingContract,
+		},
+		Message: signer.TypedDataMessage{
+			"integrationNode": math.NewHexOrDecimal256(integrationID),
+			"vehicleNode":     math.NewHexOrDecimal256(vehicleNode),
+		},
+	}
+}
+
+// GetVirtualDeviceMintingPayload godoc
+// @Description gets the payload for to mint virtual device given an integration token ID
+// @Tags        integrations
+// @Produce     json
+// @Success     200 {array} signer.TypedData
+// @Router      /integration/:tokenID/mint-virtual-device [get]
+func (vc *VirtualDeviceController) GetVirtualDeviceMintingPayload(c *fiber.Ctx) error {
+	tokenID := c.Params("tokenID")
+	vehicleNode := c.Query("vehicle_id")
+	userID := helpers.GetUserID(c)
+
+	uTokenID, err := strconv.ParseUint(tokenID, 10, 64)
+	if err != nil {
+		return fiber.NewError(fiber.StatusBadRequest, "invalid tokenID provided")
+	}
+
+	user, err := vc.usersClient.GetUser(c.Context(), &pb.GetUserRequest{
+		Id: userID,
+	})
+	if err != nil {
+		vc.log.Debug().AnErr("GetUser Grpc", err).Msg("error occurred when fetching user")
+		return fiber.NewError(fiber.StatusUnauthorized, "error occurred when fetching user")
+	}
+
+	if user.EthereumAddress == nil {
+		return fiber.NewError(fiber.StatusUnauthorized, "User does not have an Ethereum address on file.")
+	}
+
+	integration, err := vc.deviceDefSvc.GetIntegrationByTokenID(c.Context(), uTokenID)
+	if err != nil {
+		return fiber.NewError(fiber.StatusBadRequest, "failed to get integration")
```

Comment:
Translate again. Note that we're just using this to check that the integration _exists_.